ccl/backupccl: TestBackupRestoreSubsetCreatedStats failed

[(ccl/backupccl).TestBackupRestoreSubsetCreatedStats failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2180692&tab=buildLog) on [release-20.1@ba8de96cc0c40e27818194e8e548b47063ea4ae6](https://github.com/cockroachdb/cockroach/commits/ba8de96cc0c40e27818194e8e548b47063ea4ae6):

```
I200814 05:16:05.265700 234500 sql/temporary_schema.go:575  [n1] temporary object cleaner next scheduled to run at 2020-08-14 05:46:05.217928109 +0000 UTC
I200814 05:16:05.265761 234616 sql/event_log.go:132  [n1,intExec=optInToDiagnosticsStatReporting] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:root}
I200814 05:16:05.275246 233922 sql/event_log.go:132  [n1,intExec=set-setting] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:20.1 User:root}
I200814 05:16:05.288302 234993 sql/event_log.go:132  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:a1b753b5-3af5-49a7-8d79-a3c324181747 User:root}
I200814 05:16:05.295257 234910 sql/event_log.go:132  [n1,intExec=create-default-db] Event: "create_database", target: 50, info: {DatabaseName:defaultdb Statement:CREATE DATABASE IF NOT EXISTS defaultdb User:root}
I200814 05:16:05.299452 234684 sql/event_log.go:132  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
I200814 05:16:05.326846 234338 server/server.go:1798  [n1] done ensuring all necessary migrations have run
I200814 05:16:05.326890 234338 server/server.go:2040  [n1] serving sql connections
I200814 05:16:05.329305 235085 sqlmigrations/migrations.go:668  [n1] starting wait for upgrade finalization before schema change job migration
I200814 05:16:05.329554 235085 sqlmigrations/migrations.go:700  [n1] starting schema change job migration
I200814 05:16:05.332413 234338 testutils/testcluster/testcluster.go:760  WaitForFullReplication took: 637ns
I200814 05:16:05.334033 235083 server/server_update.go:55  [n1] no need to upgrade, cluster already at the newest version
I200814 05:16:05.339018 235086 sql/event_log.go:132  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:127.0.0.1:45103 Attrs: Locality:region=test,dc=dc1 ServerVersion:20.1 BuildTag:v20.1.4-50-gba8de96 StartedAt:1597382165203894472 LocalityAddress:[] ClusterName: SQLAddress:127.0.0.1:42823} ClusterID:43326666-240e-434d-9221-7de365cc45ba StartedAt:1597382165203894472 LastUp:1597382165203894472}
I200814 05:16:05.345137 235085 sqlmigrations/migrations.go:1124  [schema-change-job-migration] evaluating tables for creating jobs
I200814 05:16:05.345293 235085 sqlmigrations/migrations.go:719  [n1] schema change job migration completed
I200814 05:16:05.377042 235254 sql/event_log.go:132  [n1,client=127.0.0.1:50960,hostssl,user=root] Event: "create_database", target: 52, info: {DatabaseName:data Statement:CREATE DATABASE data User:root}
I200814 05:16:05.389114 235254 sql/event_log.go:132  [n1,client=127.0.0.1:50960,hostssl,user=root] Event: "create_table", target: 53, info: {TableName:data.public.bank Statement:CREATE TABLE IF NOT EXISTS bank (id INT8 PRIMARY KEY, balance INT8, payload STRING, FAMILY (id, balance, payload)) User:root}
I200814 05:16:05.390741 235012 kv/kvserver/replica_command.go:397  [n1,split,s1,r32/1:/{Table/36-Max}] initiating a split of this range at key /Table/53 [r33] (zone config)
I200814 05:16:05.402178 234338 workload/workloadsql/dataload.go:140  imported bank (0s, 1 rows)
I200814 05:16:05.404637 234338 testutils/testcluster/testcluster.go:760  WaitForFullReplication took: 814ns
I200814 05:16:05.408481 235254 sql/event_log.go:132  [n1,client=127.0.0.1:50960,hostssl,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:sql.stats.automatic_collection.enabled Value:false User:root}
I200814 05:16:05.413341 235254 sql/event_log.go:132  [n1,client=127.0.0.1:50960,hostssl,user=root] Event: "create_table", target: 54, info: {TableName:data.public.foo Statement:CREATE TABLE data.public.foo (a INT8) User:root}
I200814 05:16:05.415190 235593 kv/kvserver/replica_command.go:397  [n1,split,s1,r33/1:/{Table/53-Max}] initiating a split of this range at key /Table/54 [r34] (zone config)
I200814 05:16:05.431941 235637 jobs/registry.go:806  [n1] job 581015192940249089: stepping through state running with error <nil>
I200814 05:16:05.471687 235637 jobs/registry.go:806  [n1] job 581015192940249089: stepping through state succeeded with error <nil>
I200814 05:16:05.479543 235637 jobs/registry.go:961  [n1] job 581015192940249089: status succeeded after adoption finished
I200814 05:16:05.493664 235688 jobs/registry.go:806  [n1] job 581015193126895617: stepping through state running with error <nil>
I200814 05:16:05.522313 235688 jobs/registry.go:806  [n1] job 581015193126895617: stepping through state succeeded with error <nil>
I200814 05:16:05.527848 235688 jobs/registry.go:961  [n1] job 581015193126895617: status succeeded after adoption finished
I200814 05:16:05.580177 235720 jobs/registry.go:806  [n1] job 581015193377177601: stepping through state running with error <nil>
I200814 05:16:05.597380 235536 ccl/storageccl/export.go:102  [n1,s1,r33/1:/Table/5{3-4}] export [/Table/53/1,/Table/53/2)
I200814 05:16:05.597398 235535 ccl/storageccl/export.go:102  [n1,s1,r34/1:/{Table/54-Max}] export [/Table/54/1,/Table/54/2)
I200814 05:16:05.651922 235720 jobs/registry.go:806  [n1] job 581015193377177601: stepping through state succeeded with error <nil>
I200814 05:16:05.672342 235720 jobs/registry.go:961  [n1] job 581015193377177601: status succeeded after adoption finished
I200814 05:16:05.678836 235254 sql/event_log.go:132  [n1,client=127.0.0.1:50960,hostssl,user=root] Event: "create_database", target: 55, info: {DatabaseName:"data 2" Statement:CREATE DATABASE "data 2" User:root}
I200814 05:16:05.694782 235880 jobs/registry.go:806  [n1] job 581015193801228289: stepping through state running with error <nil>
I200814 05:16:05.700728 235862 kv/kvserver/replica_command.go:397  [n1,s1,r34/1:/{Table/54-Max}] initiating a split of this range at key /Table/56 [r35] (manual)
I200814 05:16:05.728144 235880 jobs/registry.go:806  [n1] job 581015193801228289: stepping through state succeeded with error <nil>
I200814 05:16:05.742431 235880 jobs/registry.go:961  [n1] job 581015193801228289: status succeeded after adoption finished
I200814 05:16:05.744279 234338 util/stop/stopper.go:539  quiescing
I200814 05:16:05.744448 235949 util/stop/stopper.go:539  quiescing
W200814 05:16:05.744867 235913 kv/kvserver/intentresolver/intent_resolver.go:745  failed to gc transaction record: could not GC completed transaction anchored at /Table/3/1/56/2/1: node unavailable; try another peer
--- FAIL: TestBackupRestoreSubsetCreatedStats (0.68s)
    backup_test.go:3822: query 'SELECT name, "columnIDs", "rowCount", "distinctCount", "nullCount" FROM system.table_statistics': expected:
        bank_stats, {1}, 1, 1, 0
        bank_stats, {2}, 1, 1, 0
        bank_stats, {3}, 1, 1, 0
        
        got:
        

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestBackupRestoreSubsetCreatedStats PKG=./pkg/ccl/backupccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestBackupRestoreSubsetCreatedStats.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/ccl/backupccl/backup_test.go
+++ b/pkg/ccl/backupccl/backup_test.go
@@ -77,6 +77,46 @@ const (
 	localFoo                    = "nodelocal://0/foo"
 )
 
+// getStatsQuery returns a SQL query that will return the properties of the
+// statistics on a table that are expected to remain the same after being
+// restored on a new cluster.
+func getStatsQuery(tableName string) string {
+	return fmt.Sprintf(`SELECT
+	  statistics_name,
+	  column_names,
+	  row_count,
+	  distinct_count,
+	  null_count
+	FROM [SHOW STATISTICS FOR TABLE %s]`, tableName)
+}
+
+// injectStats directly injects some arbitrary statistic into a given table for
+// a specified column.
+// See injectStatsWithRowCount.
+func injectStats(
+	t *testing.T, sqlDB *sqlutils.SQLRunner, tableName string, columnName string,
+) [][]string {
+	return injectStatsWithRowCount(t, sqlDB, tableName, columnName, 100 /* rowCount */)
+}
+
+// injectStatsWithRowCount directly injects some statistics specifying some row
+// count for a column in the given table.
+// N.B. This should be used in backup testing over CREATE STATISTICS since it
+// ensures that the stats cache will be up to date during a subsequent BACKUP.
+func injectStatsWithRowCount(
+	t *testing.T, sqlDB *sqlutils.SQLRunner, tableName string, columnName string, rowCount int,
+) [][]string {
+	sqlDB.Exec(t, fmt.Sprintf(`ALTER TABLE %s INJECT STATISTICS '[
+	{
+		"columns": ["%s"],
+		"created_at": "2018-01-01 1:00:00.00000+00:00",
+		"row_count": %d,
+		"distinct_count": %d
+	}
+	]'`, tableName, columnName, rowCount, rowCount))
+	return sqlDB.QueryStr(t, getStatsQuery(tableName))
+}
+
 func backupRestoreTestSetupWithParams(
 	t testing.TB,
 	clusterSize int,
@@ -3761,27 +3801,17 @@ func TestBackupCreatedStats(t *testing.T) {
 	sqlDB.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_collection.enabled=false`)
 
 	sqlDB.Exec(t, `CREATE TABLE data.foo (a INT PRIMARY KEY)`)
-	sqlDB.Exec(t, `CREATE STATISTICS foo_stats FROM data.foo`)
-	sqlDB.Exec(t, `CREATE STATISTICS bank_stats FROM data.bank`)
+	injectStats(t, sqlDB, "data.bank", "id")
+	injectStats(t, sqlDB, "data.foo", "a")
 	sqlDB.Exec(t, `BACKUP data.bank, data.foo TO $1 WITH revision_history`, localFoo)
 	sqlDB.Exec(t, `CREATE DATABASE "data 2"`)
 	sqlDB.Exec(t, `RESTORE data.bank, data.foo FROM $1 WITH skip_missing_foreign_keys, into_db = $2`,
 		localFoo, "data 2")
 
-	sqlDB.CheckQueryResults(t,
-		`SELECT statistics_name, column_names, row_count, distinct_count, null_count
-	FROM [SHOW STATISTICS FOR TABLE "data 2".bank] WHERE statistics_name='bank_stats'`,
-		[][]string{
-			{"bank_stats", "{id}", "1", "1", "0"},
-			{"bank_stats", "{balance}", "1", "1", "0"},
-			{"bank_stats", "{payload}", "1", "1", "0"},
-		})
-	sqlDB.CheckQueryResults(t,
-		`SELECT statistics_name, column_names, row_count, distinct_count, null_count
-	FROM [SHOW STATISTICS FOR TABLE "data 2".foo] WHERE statistics_name='foo_stats'`,
-		[][]string{
-			{"foo_stats", "{a}", "0", "0", "0"},
-		})
+	sqlDB.CheckQueryResults(t, getStatsQuery(`"data 2".bank`),
+		sqlDB.QueryStr(t, getStatsQuery("data.bank")))
+	sqlDB.CheckQueryResults(t, getStatsQuery(`"data 2".foo`),
+		sqlDB.QueryStr(t, getStatsQuery("data.foo")))
 }
 
 // Ensure that backing up and restoring an empty database succeeds.
@@ -3809,23 +3839,20 @@ func TestBackupRestoreSubsetCreatedStats(t *testing.T) {
 	sqlDB.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_collection.enabled=false`)
 
 	sqlDB.Exec(t, `CREATE TABLE data.foo (a INT)`)
-	sqlDB.Exec(t, `CREATE STATISTICS foo_stats FROM data.foo`)
-	sqlDB.Exec(t, `CREATE STATISTICS bank_stats FROM data.bank`)
+	bankStats := injectStats(t, sqlDB, "data.bank", "id")
+	injectStats(t, sqlDB, "data.foo", "a")
 
 	sqlDB.Exec(t, `BACKUP data.bank, data.foo TO $1 WITH revision_history`, localFoo)
-	sqlDB.Exec(t, `DELETE FROM system.table_statistics WHERE name = 'foo_stats' OR name = 'bank_stats'`)
+	// Clear the stats.
+	sqlDB.Exec(t, `DELETE FROM system.table_statistics WHERE true`)
 	sqlDB.Exec(t, `CREATE DATABASE "data 2"`)
+	sqlDB.Exec(t, `CREATE TABLE "data 2".foo (a INT)`)
 	sqlDB.Exec(t, `RESTORE data.bank FROM $1 WITH skip_missing_foreign_keys, into_db = $2`,
 		localFoo, "data 2")
 
-	// Ensure that the bank_stats have been restored, but foo_stats have not.
-	sqlDB.CheckQueryResults(t,
-		`SELECT name, "columnIDs", "rowCount", "distinctCount", "nullCount" FROM system.table_statistics`,
-		[][]string{
-			{"bank_stats", "{1}", "1", "1", "0"}, // id column
-			{"bank_stats", "{2}", "1", "1", "0"}, // balance column
-			{"bank_stats", "{3}", "1", "1", "0"}, // payload column
-		})
+	// Ensure that bank's stats have been restored, but foo's have not.
+	sqlDB.CheckQueryResults(t, getStatsQuery(`"data 2".bank`), bankStats)
+	sqlDB.CheckQueryResults(t, getStatsQuery(`"data 2".foo`), [][]string{})
 }
 
 // Ensure that statistics are restored from correct backup.
@@ -3841,19 +3868,18 @@ func TestBackupCreatedStatsFromIncrementalBackup(t *testing.T) {
 
 	sqlDB.Exec(t, `SET CLUSTER SETTING sql.stats.automatic_collection.enabled=false`)
 
-	// Create the 1st backup, where data.bank has 1 account.
-	sqlDB.Exec(t, `CREATE STATISTICS bank_stats FROM data.bank`)
+	// Create the 1st backup, with stats estimating 50 rows.
+	injectStatsWithRowCount(t, sqlDB, "data.bank", "id", 50 /* rowCount */)
 	sqlDB.Exec(t, `BACKUP data.bank TO $1 WITH revision_history`, localFoo)
 
-	// Create the 2nd backup, where data.bank has 3 accounts.
-	sqlDB.Exec(t, `INSERT INTO data.bank VALUES (2, 2), (4, 4)`)
-	sqlDB.Exec(t, `CREATE STATISTICS bank_stats FROM data.bank`)
+	// Create the 2nd backup, with stats estimating 100 rows.
+	injectStatsWithRowCount(t, sqlDB, "data.bank", "id", 100 /* rowCount */)
+	statsBackup2 := sqlDB.QueryStr(t, getStatsQuery("data.bank"))
 	sqlDB.QueryRow(t, `SELECT cluster_logical_timestamp()`).Scan(&beforeTs) // Save time to restore to this point.
 	sqlDB.Exec(t, `BACKUP data.bank TO $1 INCREMENTAL FROM $2 WITH revision_history`, incremental1Foo, localFoo)
 
-	// Create the 3rd backup, where data.bank has 5 accounts.
-	sqlDB.Exec(t, `INSERT INTO data.bank VALUES (3, 3), (5, 2)`)
-	sqlDB.Exec(t, `CREATE STATISTICS bank_stats FROM data.bank`)
+	// Create the 3rd backup, with stats estimating 500 rows.
+	injectStatsWithRowCount(t, sqlDB, "data.bank", "id", 500 /* rowCount */)
 	sqlDB.Exec(t, `BACKUP data.bank TO $1 INCREMENTAL FROM $2, $3 WITH revision_history`, incremental2Foo, localFoo, incremental1Foo)
 
 	// Restore the 2nd backup.
@@ -3861,17 +3887,8 @@ func TestBackupCreatedStatsFromIncrementalBackup(t *testing.T) {
 	sqlDB.Exec(t, fmt.Sprintf(`RESTORE data.bank FROM "%s", "%s", "%s" AS OF SYSTEM TIME %s WITH skip_missing_foreign_keys, into_db = "%s"`,
 		localFoo, incremental1Foo, incremental2Foo, beforeTs, "data 2"))
 
-	// Expect the values in row_count and distinct_count to be 3. The values
-	// would be 1 if the stats from the full backup were restored and 5 if
-	// the stats from the latest incremental backup were restored.
-	sqlDB.CheckQueryResults(t,
-		`SELECT statistics_name, column_names, row_count, distinct_count, null_count
-	FROM [SHOW STATISTICS FOR TABLE "data 2".bank] WHERE statistics_name='bank_stats'`,
-		[][]string{
-			{"bank_stats", "{id}", "3", "3", "0"},
-			{"bank_stats", "{balance}", "3", "3", "0"},
-			{"bank_stats", "{payload}", "3", "2", "2"},
-		})
+	// Expect the stats look as they did in the second backup.
+	sqlDB.CheckQueryResults(t, getStatsQuery(`"data 2".bank`), statsBackup2)
 }
 
 // TestProtectedTimestampsDuringBackup ensures that the timestamp at which a

--- a/pkg/ccl/backupccl/full_cluster_backup_restore_test.go
+++ b/pkg/ccl/backupccl/full_cluster_backup_restore_test.go
@@ -116,7 +116,7 @@ func TestFullClusterBackup(t *testing.T) {
 	sqlDB.Exec(t, `GRANT CREATE, SELECT ON DATABASE data TO system_ops;`)
 	sqlDB.Exec(t, `GRANT system_ops TO maxroach1;`)
 
-	sqlDB.Exec(t, `CREATE STATISTICS my_stats FROM data.bank`)
+	injectStats(t, sqlDB, "data.bank", "id")
 	sqlDB.Exec(t, `BACKUP TO $1`, localFoo)
 
 	// Create a bunch of user tables on the restoring cluster that we're going

